File: test/integration-tests/sdk-test-suites/cpp/function_provider.cpp
/*****************************************************************************
 * Copyright (c) 2021 Bosch.IO GmbH
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 ****************************************************************************/

#include <csignal>
#include <fstream>
#include <iostream>
#include <memory>

#include "nlohmann/json.hpp"
#include "client.hpp"

using json = nlohmann::json;

using iotea::core::Client;
using iotea::core::FunctionTalent;
using iotea::core::ProtocolGateway;
using iotea::core::call_ctx_ptr;

static const char TALENT_ID[] = "functionProvider-cpp";
static const char FUNC_ECHO[] = "echo";

class FunctionProvider : public FunctionTalent {
   public:
    FunctionProvider()
        : FunctionTalent(TALENT_ID) {
        RegisterFunction(FUNC_ECHO, [](const json& args, call_ctx_ptr context) {
            context->Reply(args[0]);
        });
    }
};

std::shared_ptr<Client> client;

void signal_handler(int) {
    if (client) {
        client->Stop();
    }
}

int main(int, char**) {
    std::ifstream file{"../../../config/tests/cpp/config.json"};
    std::string config{std::istreambuf_iterator<char>(file), std::istreambuf_iterator<char>()};

    auto gateway = std::make_shared<ProtocolGateway>(json::parse(config));
    client = std::make_shared<Client>(gateway);

    auto talent = std::make_shared<FunctionProvider>();
    client->RegisterFunctionTalent(talent);

    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    client->Start();

    return 0;
}

File: test/integration-tests/sdk-test-suites/javascript/functionProvider.js
//##############################################################################
// Copyright (c) 2021 Bosch.IO GmbH
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0
//##############################################################################
// uncomment for local developer setup
// const iotea = require('../../../../src/module.js');
const iotea = require('boschio.iotea');

const {
    FunctionTalent
} = iotea;

const {
    Logger,
    JsonModel
} = iotea.util;

const config = new JsonModel(require('../../config/tests/javascript/config.json'));
process.env.LOG_LEVEL = config.get('loglevel', Logger.ENV_LOG_LEVEL.INFO);

class FunctionProvider extends FunctionTalent {
    constructor(protocolGatewayConfig) {
        super('function-provider-js', protocolGatewayConfig);

        // Register Functions
        this.registerFunction('echo', this.echo.bind(this));
    }

    async echo(value, ev, evtctx, timeoutAtMs) {
        this.logger.debug('Echo called');
        return value;
    }
}

const pgConfig = config.get("protocolGateway")
const fp = new FunctionProvider(pgConfig);

fp.start();
